### PR TITLE
1973. Validation of personal details pages.

### PR DESCRIPTION
**Pull request by Itay-Kirsh**
https://tools.hmcts.net/jira/browse/RET-1973

Added missing tests for RET-1973.
Note, accidentally merged 1973 into master before QA approval.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by brendanmchugh — APPROVED** [review]

**Review by marcjudge — APPROVED** [review]

**Review by brendanmchugh — DISMISSED** [review]

File: src/main/definitions/case.ts [hunk 1]
```diff
@@ -92,6 +92,10 @@ export interface Case {
   claimantPensionWeeklyContribution?: number;
   reasonableAdjustments?: YesOrNo;
   reasonableAdjustmentsDetail?: string;
+  gender?: string;
```
Comment (marcjudge) [inline src/main/definitions/case.ts:95]: Let's try to keep these names consistent with the ccd definitions file to minimise confusion. Properties should be called 'claimant_sex', 'claimant_gender_identity', 'claimant_title1', 'claimant_title_other'. See here - https://github.com/hmcts/et-ccd-definitions-englandwales/blob/master/definitions/json/ComplexTypes.json

**Review by brendanmchugh — DISMISSED** [review]

**Review by marcjudge — DISMISSED** [review]

File: src/main/components/form/dateValidators.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+  const enteredDate = new Date(+date.year, +date.month - 1, +date.day);
+  const datePlus10 = new Date();
+  datePlus10.setFullYear(datePlus10.getFullYear() + 10);
+
+  if (enteredDate > datePlus10) {
+    return { error: 'invalidDateMoreThanTenYearsInFuture', fieldName: 'year' };
+  }
+};
+
+/**
```
Comment (JackReevies) [inline src/main/components/form/dateValidators.ts:150]: Not sure what our rules are for JSDocs but why do some methods have them and others not?

File: src/main/components/form/dateValidators.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+ * Comparison of dates excluding specific hours/minutes/seconds of the day.
+ */
+export const isFirstDateBeforeSecond = (value1: CaseDate, value2: CaseDate): boolean => {
+  return (
+    new Date(+value1.year, +value1.month - 1, +value1.day) < new Date(+value2.year, +value2.month - 1, +value2.day)
+  );
+};
+
+export const isDateEmpty = (date: CaseDate): boolean => {
+  return date === undefined || Object.values(date).every(e => e === undefined || e === '');
```
Comment (JackReevies) [inline src/main/components/form/dateValidators.ts:186]: return `!date || Object.values(date).every(e => !e)` ?

File: src/main/components/form/validator.ts [hunk 4]
```diff
@@ -335,20 +180,10 @@ export const isValidCurrency: Validator = value => {
...
-export const validatePreferredOther: Validator = value => {
-  let outcome: string = undefined;
-  if (otherGenderTitle === 'Other') {
-    if ((value as string).trim().length === 0) {
-      outcome = 'required';
-    } else if (/^\d+$/.test(value as string) || /^\s*\d/.test(value as string)) {
-      outcome = 'numberError';
-    }
-    return outcome;
+export const validateTitlePreference: Validator = value => {
```
Comment (JackReevies) [inline src/main/components/form/validator.ts:183]: Would it work to declare value as a string in the signature? (ie ` = (value: string) => {`). Would eliminate needing to "cast" it multiple times in the body

File: src/main/definitions/dates.ts [hunk 5]
```diff
@@ -96,5 +105,9 @@ export const StartDateFormFields = {
   hint: (l: AnyRecord): string => l.hint,
   values: DateValues,
   validator: (value: CaseDate): DateTypes =>
-    areDateFieldsFilledIn(value) || isDateInputInvalid(value) || isFutureDate(value),
+    areDateFieldsFilledIn(value) || isDateInputInvalid(value) || isDateInPast(value),
+};
+
+export const convertDateToCaseDate = (date: Date): CaseDate => {
+  return { day: `${date.getDate()}`, month: `${date.getMonth() + 1}`, year: `${date.getFullYear()}` };
```
Comment (JackReevies) [inline src/main/definitions/dates.ts:112]: these don't need to be in template literals

File: src/main/helper/ApiFormatter.ts [hunk 6]
```diff
@@ -104,11 +105,23 @@ export function toApiFormat(caseItem: CaseWithId): UpdateCaseBody {
...
+  return (
+    dobDate.year +
+    '-' +
+    (dobDate.month.match(/^\d$/) ? '0' : '') +
+    dobDate.month +
+    '-' +
+    (dobDate.day.match(/^\d$/) ? '0' : '') +
+    dobDate.day
+  );
+};
```
Comment (JackReevies) [inline src/main/helper/ApiFormatter.ts:122]: Template literal would read nicer here, also if the intention of the match is to add a preceding 0. you can use `padStart(2, '0')` to make sure its length is always 2. (ie, `dobDate.month.padStart(2, '0')`)

```return `${dobDate.year}-${dobDate.month.padStart(2, '0')}-${dobDate.day.padStart(2, '0')}` ```

**Review by JackReevies — CHANGES_REQUESTED** [review]
Looks good, just a couple of syntax queries

File: src/main/definitions/dates.ts (hunk 5 above)
Comment (Itay-Kirsh) [reply src/main/definitions/dates.ts:112]: I needed to convert them to strings anyway (that's how CaseData stores it), is there a nicer way to do it?

File: src/main/components/form/dateValidators.ts (hunk 2 above)
Comment (Itay-Kirsh) [reply src/main/components/form/dateValidators.ts:150]: I see that not many places have them, I'll try and make the method names a lot clearer.

File: src/main/components/form/dateValidators.ts [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+export type DateValidator = (
+  value: CaseDate | undefined,
+  value2?: CaseDate | undefined
+) => void | string | InvalidField;
+
+export const areDateFieldsFilledIn: DateValidator = (fields: CaseDate | undefined) => {
+  if (
+    typeof fields !== 'object' ||
+    Object.keys(fields).length !== 3 ||
+    Object.values(fields).every(e => e === null || e === '')
```
[the remark this answers is not in the document]
Comment (Itay-Kirsh) [reply src/main/components/form/dateValidators.ts:16]: Using isDateEmpty

File: src/main/components/form/dateValidators.ts [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+      };
+    }
+  }
+};
+
+export const isDateNotPartial: DateValidator = (date: CaseDate | undefined) => {
+  if (date === undefined || isDateEmpty(date) || !Object.values(date).some(e => e === null || e === '')) {
+    return;
+  }
+  for (const [fieldName, field] of Object.entries(date)) {
```
[the remark this answers is not in the document]
Comment (Itay-Kirsh) [reply src/main/components/form/dateValidators.ts:37]: Deduplicated

File: src/main/components/form/dateValidators.ts (hunk 3 above)
Comment (Itay-Kirsh) [reply src/main/components/form/dateValidators.ts:186]: Cleaned

File: src/main/components/form/validator.ts (hunk 4 above)
Comment (Itay-Kirsh) [reply src/main/components/form/validator.ts:183]: That would be much a larger change and not as part of this commit, I've only just added another Validator with its existing signature.

**Review by JackReevies — DISMISSED** [review]

**Review by robin-ieong — APPROVED** [review]

**Review by robin-ieong — DISMISSED** [review]

**Review by JackReevies — DISMISSED** [review]

File: src/main/helper/ApiFormatter.ts [hunk 9]
```diff
@@ -112,20 +113,20 @@ export function toApiFormat(caseItem: CaseWithId): UpdateCaseBody {
...
 
-function parseDateFromString(dobDate: string): CaseDate {
-  if (dobDate) {
-    const year = dobDate.substring(0, 4);
-    const month = dobDate.substring(5, 7);
-    const day = dobDate.substring(8);
+  return `${date.year}-${date.month.padStart(2, '0')}-${date.day.padStart(2, '0')}`;
+};
 
+export const parseDateFromString = (dobDate: string): CaseDate => {
```
Comment (marcjudge) [inline src/main/helper/ApiFormatter.ts:124]: Could you please also rename this parameter to date?

**Review by lvo700 — APPROVED** [review]

**Review by marcjudge — APPROVED** [review]

**Pull request merged**